Use pre-built images for nginx & service containers

alxndrsn opened the pull request:
# TODO

- [ ] review tag script - is it necessary?
- [ ] consider how commits for upgrading submodules are currently handled - tagging could likely be identical

---

Related to #677 

# What's changed?

* `docker-compose.yml` now references `service` and `nginx` containers by specific version number
* add a script for tagging releases and making appropriate updates to `docker-compose.yml`.  If preferred, this could be converted to documentation

-----

#### What has been done to verify that this works as intended?

- [ ] CI
- [ ] a lot of manual testing in a forked repo (https://github.com/alxndrsn/odk-central/actions)
- [ ] ran `docker compose build` locally, and watched the results:

#### Why is this the best possible solution? Were any other approaches considered?

TODO

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

* should reduce the system requirements for performing upgrades
* might contain surprises for deployers who have made local changes to any of client, server, nginx etc.

#### Does this change require updates to documentation? If so, please file an issue [here](https://github.com/getodk/docs/issues/new) and include the link below.

- [ ] internal release documentation
- [ ] public upgrade documentation

#### Before submitting this PR, please make sure you have:

- [x] branched off and targeted the `next` branch OR only changed documentation/infrastructure (`master` is stable and used in production)
- [x] verified that any code or assets from external sources are properly credited in comments or that everything is internally sourced

Commit:
ci: move to GHA; only run ghcr build on test pass AND when tagged

Commit:
fix shellcheck?

Commit:
fix docker-compose build context

Commit:
volume paths

Commit:
ci: comment out working stuff

Commit:
service.dockerfile: separate unrelated commands

Commit:
fix runner name

Commit:
check path with available tools

Commit:
1..timeout

Commit:
for inc loop

Commit:
store result before testing

Commit:
less subprocess

Commit:
store res correct

Commit:
remove circle config

Commit:
get container name

Commit:
fix syntax hilighting

Commit:
no container: prefix

Commit:
start container before getting name

Commit:
rename var

Commit:
exec

Commit:
add logging

Commit:
remove noisy x

Commit:
restructure

Commit:
update check regex

Commit:
clearer code

Commit:
remove nginxContainer var

Commit:
test images

Commit:
add docker compose logs

Commit:
X

Commit:
or true?

Commit:
use a temp file

Commit:
single line

Commit:
redirect to stdout

Commit:
run it

Commit:
log spacing

Commit:
show pm2 before count

Commit:
8

Commit:
expect 4

Commit:
enable all jobs

Commit:
clearer logging, comments

Commit:
docker-compose: use published images

Commit:
use snapshots but not for root docker-compos

Commit:
use both docker-compose files

Commit:
ci: add TODO

Commit:
fix compose file order

Commit:
fix context

Commit:
remove explicit context

Commit:
ghcr: show expected tag

Commit:
run ghcr job

Commit:
docker run

Commit:
try execute?

Commit:
try wider conditionsal

Commit:
comment

Commit:
comment out test yml

Commit:
print event name

Commit:
move test yml out the way

Commit:
fix yml

Commit:
more devbug - tags etc

Commit:
remove TODO

Commit:
no need

Commit:
ref type

Commit:
try check only on tag

Commit:
ref_type

Commit:
simplify if the second half runs

Commit:
reinstate tings

Commit:
reinstate test

Commit:
revert dockerfile changes

Commit:
don't build on master

Commit:
run test on tag

Commit:
expicitly run on tags

Commit:
more restruct

alxndrsn committed:
Testing build triggers

Merged to see if this affects workflow behaviours

Commit:
rebuild

Commit:
Merge branch 'next' into another-try

Commit:
build on pushes (check it still builds on tags...)

Commit:
simplify/revert yaml

alxndrsn commented on rebuild:
Make sure this file is removed before merge.

Commit:
Add debug job

Commit:
fix yaml?

Commit:
stub out test jobs

alxndrsn commented on .github/workflows/test.yml:
```diff
@@ -5,46 +5,7 @@ on:
 jobs:
-  test-misc: # quick, simple checks
-    timeout-minutes: 2
+  dummy:
```
reinstate normal jobs before review

Commit:
remove rebuild

Commit:
Add make release target

Commit:
Revert "Add make release target"

This reverts commit 1fc807beb675b4349cc03a3fa8c7f6ebf196d893.

Commit:
import ghcr into test

Commit:
remove ghcr workflow

Commit:
re-intro standard CI

Commit:
remove extra permissiosn

Commit:
add tag check

Commit:
add release script

Commit:
add release script

Commit:
tweak script

Commit:
shellcheck

Commit:
ci: rename check-image-references job

lognaturel commented on docker-compose.yml:
```diff
@@ -34,9 +34,7 @@ services:
-    build:
-      context: .
-      dockerfile: service.dockerfile
+    image: 'ghcr.io/getodk/central-service:2025.1'
```
We'd want this to start with a known version so e.g. https://github.com/getodk/central/pkgs/container/central-service/335981709?tag=v2024.3.1 I think

lognaturel commented:
Nice! We can automate substituting in the tagged version in the docker compose file and then tag that commit. 👍 

@matthew-white what do you think about running a script to release? The tag would then show up in Github and we could write the release notes there as usual.

I thought briefly about moving that functionality to the release Github action but that won't quite work -- we want to tag the commit that makes the change to the yml file.

I think we'll want to add some documentation on how to make code edits to frontend and backend since that will now require changes to the docker compose file as well. It's not something we highly recommend but it was easy in the past and we know some people do it.

We'll need to build from local source for the staging server. Maybe we could consider conditionally building or using an image based on a setting (could be a follow-up PR)?

matthew-white commented:
I'd be OK with running a script. :+1: However, I don't understand yet the intended release process as a whole. Very possibly I'm just missing something. From the script, it looks like we'd run the script on the `master` branch of the `central` repo. But wouldn't we need to run the script on `next` before merging `next` into `master`? Otherwise, there would be a commit on the `master` branch that includes changes to the `central` repo without corresponding changes to `central-backend` and `central-frontend`. For example, updating Node requires changing `central` and sometimes also requires code changes to `central-backend`. If we merged that change to `central` into `master` and only then ran the script, then there'd be a commit and a short time window in which `master` doesn't contain working code. In other words, I'm wondering about the chance that a user could install Central at an awkward time between when the `next` branch is merged and the script is run. Is there a way to keep the release fully atomic? Could the script be run on the `next` branch instead of `master`? Or could the script merge `next` into `master` as part of its process?

matthew-white commented:
> We'll need to build from local source for the staging server.

I think this is also something that we'd need to do on the QA/test server during regression testing.

alxndrsn committed:
Merge branch 'next' into another-try

Commit:
use existing version

alxndrsn committed:
Merge branch 'next' into another-try

matthew-white commented:
Sort of a random thought / only somewhat related, but if we're going to have a script for the release, could one thing that the script does be generating version.txt? Like could we remove that generation from nginx.dockerfile? If so, then maybe we wouldn't need to include the .git/ directory in the Docker context.

lognaturel commented:
> could one thing that the script does be generating version.txt?

We also want to be able to use images for non-released builds so I think this is a possible idea but it would have to be part of the CI-based image building, not specifically part of the tagging/release script.

lognaturel commented:
[Latest thinking doc](https://docs.google.com/document/d/1ePncwTY6Zl9M7Y0KtKULzFw_hCkpc0dl9i3RFVIduNQ/edit?tab=t.0#heading=h.gx9w2j9hzvnc). @alxndrsn will revise this PR in light of those thoughts and describe how each end-user goal is met by the proposed approach.

Commit:
Merge branch 'next' into another-try

Commit:
fix release shebang

Commit:
remove ghcr workflow again

Commit:
tag-release.sh: update to block running on `master` branch

alxndrsn commented:
> But wouldn't we need to run the script on `next` before merging `next` into `master`? Otherwise, there would be a commit on the `master` branch that includes changes to the `central` repo without corresponding changes to `central-backend` and `central-frontend`.

I think this is correct.  To tackle this I have:

1. renamed `release.sh` to `tag-release.sh`, and
2. changed the branch check from *only* runs on `master` to *cannot* run on `master`

Commit:
Add script compose-snapshots.sh

alxndrsn commented:
> I think we'll want to add some documentation on how to make code edits to frontend and backend since that will now require changes to the docker compose file as well. It's not something we highly recommend but it was easy in the past and we know some people do it.

Is there any risk of breaking these peoples' existing deployments?  Would it be better for existing commands to continue building locally, and add new recommended commands which use pre-built images?

lognaturel commented:
> Is there any risk of breaking these peoples' existing deployments?

It would definitely be annoying for them. But if they're modifying source we know they have technical capacity and as long as we document a process I think it's ok.

> Would it be better for existing commands to continue building locally, and add new recommended commands which use pre-built images?

Can you elaborate a bit on what that would look like? When you say commands, do you mean `docker compose up`, etc?

alxndrsn commented:
> When you say commands, do you mean `docker compose up`, etc?

Yes, so in this scenario the existing `docker compose up` etc. would still work for deployments building locally, and then documentation would change for new deployments to use pre-built images, perhaps with dedicated scripts e.g. `./restart-services.sh`.

I don't think this necessarily simplifies things though, so I'm relieved you said this:

> as long as we document a process I think it's ok.

matthew-white commented:
@alxndrsn, just wanted to check whether there are any outstanding questions here. Are you blocked by anything? Once we get v2025.1 out the door, I'd be happy to meet/pair on this PR. I'm familiar with our release process(es) and how the different repos work, so I think I can add any context that's missing.

alxndrsn committed:
Merge branch 'next' into another-try

Commit:
Merge branch 'next' into another-try

Commit:
build all images

Commit:
build on tags

Commit:
Release version: v0.0.1-testing-1

Commit:
include all in tag

Commit:
Release version: v0.0.1-testing-2

Commit:
update test-secrets

Commit:
rename

Commit:
relax

Commit:
dash

Commit:
Release version: v0.0.1-testing-3

Commit:
try using bake + docker-compose?

alxndrsn committed:
Merge branch 'next' into another-try

alxndrsn committed:
Merge branch 'next' into another-try
